prusti-specs/src/lib.rs

```diff
@@ -421,3 +421,37 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
         _ => { unimplemented!() }
     }
 }
+
+pub fn predicate(attr: TokenStream, tokens: TokenStream) -> TokenStream {
+    if !attr.is_empty() {
+        // TODO: what here? error? re-emit the previous attributes? not sure
+        //       what the semantics for that are supposed to be..
+    }
+
+    let item: syn::Item = handle_result!(syn::parse2(tokens));
+    let item_span = item.span();
+    match item {
+        syn::Item::Fn(pred_func) => {
+            let body_tokens = pred_func.block.to_token_stream();
+            // We've got a function, we know the body must be a braced group.
+            let assertion_tokens =if let Some(TokenTree::Group(group)) = body_tokens.into_iter().next() {
+                group.stream()
+            } else {
+                unreachable!("A function's block must be a brace-delimited `TokenTree::Group`")
+            };
+
+            let sig = pred_func.sig.to_token_stream();
+            parse_quote_spanned! {item_span =>
+                #[requires(#assertion_tokens)]
+                #[pure]
+                // TODO: is `true` the only possibility here? do we require
+                //       all predicates to be `-> bool`?
```

Comment:
Is the comment is ignored by `parse_quote_spanned`? If not or if unsure, place it before the macro call. Anyway, I think it's correct to only have boolean predicates.

Reply:
Comments are not part of the AST (`syn`'s or Rust's), so these are definitely stripped during lexing of `lib.rs`.